docs/proposal-partially-satisfied-prerequisites.md

```diff
@@ -0,0 +1,347 @@
+
+Consider the following workflow:
+
+```ini
+[scheduling]
+    [[graph]]
+        R1 = """
+             foo ?=> bar => qux
+             foo:fail ?=> baz => qux
+             """
+[runtime]
+    [[foo,bar,baz,qux]]
+        script = true
+```
+
+The user intended `qux` to run but has made a mistake in the graph (should have been `bar | baz => qux`).
+This workflow will now stall due to `qux` being in the `waiting` state with nothing left to run.
+
+Another example:
+
+```ini
+[scheduling]
+    cycling mode = integer
+    initial cycle point = 1
+    final cycle point = 10
+    [[graph]]
+        P1 = """
+             model[-P1] => model => archive
+             archive[-P1] => archive
+             archive:fail ?=> recover
+             """
+[runtime]
+    [[model,archive,recover]]
+        script = sleep 2
+    [[archive]]
+        pre-script = [[ $CYLC_TASK_CYCLE_POINT != 2 ]] || exit 1
+```
+
+A simple workflow where a recovery task has been added to deal with the rare case where the archiving fails.
+However, a mistake in the graph means the prerequisites for `archive` can't be met in the cycle after a failure (should have been `archive[-P1] | recover[-P1] => archive`).
+In this case `archive` will becomed "blocked" and the workflow will eventually stall due to hitting the runahead limit.
+
+Similar situations could occur operationally when they have to intervene due to problems.
+Some tasks might get skipped in one cycle which could then lead to partially satisfied prerequisites in the next cycle.
+Reload is another potential cause for issues.
+Graphing changes to a running workflow could easily result in inconsistencies which cause partially satisfied prerequisites in the initial cycles.
+However, any "required" tasks will now get "blocked" and/or cause a stall so will not go unnoticed.
+
+Re-flow is also a potential cause for partially satisfied prerequisites.
+For example:
+
+```ini
+[scheduling]
+    cycling mode = integer
+    initial cycle point = 1
+    [[graph]]
+        P1 = """
+             bar[-P1] => foo => bar & baz
+             baz[-P1] => baz
+             """
+```
+
+If you run this with `cylc play --starttask=bar.2` then `baz.3` will remain stuck `waiting` and the workflow will eventually stall due to hitting the runahead limit.
+
+The new trigger syntax also deals with this example:
+
+```
+        a | r1 => b ?=> r2
+        a:fail ?=> r1 => r2
+```
+
+In this case, recovering from the failure of `a` involves running `r1` before running `b` and then running `r2` once `b1` has succeeded.
+If `a` succeeds then `r2` will have partially satisified dependencies but this is not a problem since it is optionally triggered.
+Note that this is an example of where a task is optionally triggered by one task but required by another.
+This is not a problem.
+`r2` only becomes required if `r1` is run.
+
```

Comment:
Response to this example for the record. 

This is actually a pathological graph for Cylc 7 and 8: it requires the same tasks to have different dependencies on alternate branches, but our graph parsing assumes all instances of a given task name in the graph refer to the same task, and we give the task definition all of the harvested dependencies.

We intend to fix this in Cylc 9 with dynamic subgraphs:
```
a:succeed => {r1 => b => r2}
a:fail => {r1 => r2}
```
In the meantime, for Cylc 7 and 8 we have to use different logical tasks to represent the same job content on different branches.

Also, the proposed trigger syntax is hard to interpret here. Counter-intuitively the optional trigger `b ?=> r2` does not correspond to an optional output of `b`. Instead it means *"If b runs we will spawn r2, but r2 can only run if it also gets spawned by other non-optional triggers"*.  Worse, the other non-optional trigger is an artificial dependency that won't
come naturally to users: `r1 => r2` (the natural chain is `r1 => b => r2`).

That strikes me as a convoluted bodge that works for the wrong reasons. We'd be using optional and artificial triggers to get around the nasty fact that `b` has accumulated all the triggers from two different mutually exclusive branches.
So I wouldn't be surprised if we could dream up cases that fail the bodge.

Also, we'd end up with a partially satisfied waiting `r2` every time the usual success branch runs, and I'm not entirely convinced we could prune that immediately if the unsatisfied prerequisite is marked "optional". What if the
two branches aren't mutually exclusive? The optional trigger might still get satisfied later on...